### PR TITLE
port to dune

### PR DESCRIPTION
**Pull request by Vighnesh-V**
This PR ports the entire project to dune. We can now take advantage of incremental builds + build watching :D and dependency management. Dune also generates appropriate merlin files and documentation for us.

Additionally, there are changes to modularity with reporting and grading - these are not abstracted away, and accessed via command line options, and are ready to be tweaked and implemented in a full way. For example, we only had student reporting (the old reporting format including that included the raw code, the warning, and the improvement), but now we also have ta reporting, a more compact form that reports line numbers + warnings, as well as the final score.

Finally, the linter has moved to a much easier drop-in format, at the expense of simplicity. We now use the raw Parsetree.expression type everywhere. Adding a new rule, simply requires changes in three places - 1) adding the rule in lib/checkers/, 2) two allowing the rule to be exposed in lib/checkers/checkers.ml, and 3) potentially modifiying the finder expression in lib/traverse/find.ml, as opposed to the old workflow.

### CODE REVIEW

File: astutils.opam [hunk 1]
```diff
@@ -0,0 +1,19 @@
+# This file is generated by dune, edit dune-project instead
```
[the remark this answers is not in the document]
Comment (KeenWill) [reply astutils.opam:1]: also same question for the other autogenerated files ofc

File: README.md [hunk 2]
```diff
@@ -1,20 +1,24 @@
...
 ### Note:
 This project is dependent on compiler-libs, an inherently unstable library that
 changes between OCaml installations.
 
-### Overview of command line options
-`make` to build `camelot`. This produces the `camelot` binary that can be run
-through the command line.
+### Build and run instructions
+Build:
+`dune build bin/camelot.exe`
```
[the remark this answers is not in the document]
Comment (Vighnesh-V) [reply README.md:18]: linux biased even though it says exe. Ocaml for windows is basically non existent lol

[the remark this answers is not in the document]
Comment (Vighnesh-V) [reply README.md:18]: it'll run fine on a macos / linux machine - and in theory work for windows builds too

File: astutils.opam (hunk 1 above)
[the remark this answers is not in the document]
Comment (Vighnesh-V) [reply astutils.opam:1]: this should be there. It's not actually autogenerated, but dune needs it for opam integration. All the dune files aren't auto generated - they're hand written and also need to be there. The only files that shouldn't be there are those that start with .#filename.ml, .merlin, and _build.

But yeah - it's a one time generated file that should be there.

File: README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (KeenWill) [reply README.md:18]: cool

File: astutils.opam (hunk 1 above)
[the remark this answers is not in the document]
Comment (KeenWill) [reply astutils.opam:1]: sounds good

File: README.md [hunk 3]
```diff
@@ -1,20 +1,24 @@
...
 
 ## Dependencies 
-Install the following with `opam install` after creating a switch:
+- dune (v 2.4.0) - dune will manage installing the appropriate dependencies for you I believe
+- ocamlc v4.09.0
 - ocamlfind (for debugging / dev)
-- ocamlbuild
-- ppx_tools ( for debugging / dev )
-- compiler-libs
+- compiler-libs.common
```
Comment (KeenWill) [inline README.md:9]:
```suggestion
- compiler-libs.common
- fswatch (for Build + Watch)
```

File: README.md [hunk 4]
```diff
@@ -1,20 +1,24 @@
...
+Build:
+`dune build bin/camelot.exe`
 
-For dev purposes the `make test` rule is setup so that you can run the binary on files in a particular directory, as specified in the `TESTS` variable.
\ No newline at end of file
+Build + Watch:
+`dune build bin/camelot.exe -w`
+
+Run 'tests':
+`dune exec -- bin/camelot.exe <camelot args here>`
```
Comment (KeenWill) [inline README.md:24]: Doesn't necessarily have to happen in this PR, but we should document the camelot args.

**Pull request merged**